### ISSUE TITLE
Trend Micro 11 kills the processes running in mounted devices (since 1.0.0-RC4)

### ISSUE THREAD

**Issue by sdecugis** [dokan-dev/dokany#360]
# Environment
 * Windows version:  Win7 64 bits
 * Processor architecture:  AMD64
 * Dokany version: 1.0.0 (seen also with 1.0.0-RC4, not seen with 1.0.0-RC3)
 * Library type (Dokany/FUSE): Dokany
 
# Check List
  - [v ] I checked my issue [doesn't exist yet](https://github.com/dokan-dev/dokany/issues?utf8=%E2%9C%93&q=is%3Aissue)
  - [v ] My issue is valid with mirror default sample and not specific to my user-mode driver implementation
  - [v ] I can always reproduce the issue with the provided description below.
  - [v ] I have updated Dokany to the latest version (https://github.com/dokan-dev/dokany/releases) and have reboot my computer after.

# Description

When a specific application runs from within a device mounted with Dokany 1.0.0, the application is killed by the system and the device is unmounted (mirror.exe exits by itself). After uninstalling Trend Micro 11 from this machine, the application runs normally under the mounted device.

Same behavior is observed when this application runs in the device mounted with "mirror.exe", but the attached logs are provided from our custom user-space application.

This abnormal behavior is not observed when using Dokany 1.0.0-RC3 on this machine with the same antivirus, but it is observed with Dokany 1.0.0-rc4.

This behavior is not observed with Trend Micro A/V 10.x release. 

If using the "mount manager" option, the system freezes completely at the time where the application and mirror are killed by the system when the option is not specified.

I had a look at the kernel logs, it seems that the Trend Micro AV process (TMBSRV.exe) is trying to open the file of the running application, but the user-space callback is never called. I have no idea why.

# Logs

[pid_map.txt](https://github.com/dokan-dev/dokany/files/505433/pid_map.txt)
[userspace_log.txt](https://github.com/dokan-dev/dokany/files/505434/userspace_log.txt)
[kernel_driver_log.txt](https://github.com/dokan-dev/dokany/files/505435/kernel_driver_log.txt)

**Comment by Liryna** [comment]
Hi @sdecugis ,

I have install Trend Micro A/V on my VM (http://trial.trendmicro.com/US/TM/2017/TrendMicro_TAV_11.0_US-en_Downloader.exe)

I can run the mirror but it is extremly slow, is it also the case for you ? (didn't start an application from it for now).

**Comment by sdecugis** [comment]
Hi @Liryna, thanks for the fast action.

I did not see that is was particularly slow, but it is happening on a fairly powerful machine, and the issue happens within a few seconds of starting so I may not notice lower performance...

**Comment by Liryna** [comment]
OK :D

Do you think I can reproduce the behavior with a simple Hello World print basic.exe ?

**Comment by sdecugis** [comment]
my CONTENT.exe is doing a number of things before the issue happens. The last thing I can see relates to trying to open a DLL file that does not exists (HID.dll that you can see in the log). So a simple "Hello World.exe" might be too short execution time for the issue to happen... Maybe you can try "Hello World Sleep 5000.exe" ? :)

**Comment by Liryna** [comment]
I tried to launch procmon.exe from the mirror by command line (using explorer totally slow down the VM).
It began to start annnd it became slow could not go futher...but the device was still here :cry: 

Probably should first see why this so slow with this AV.

**Comment by marinkobabic** [comment]
The reason of unmounting is the timeout. How many threads did you specify when mounting the dokan drive? If you mount the drive, can you navigate using explorer before starting the content.exe? It seems like all the threads are busy  or the usermode requests can't reach the driver for some reason. This is the reason why the device is unmounted. Kernel thinks that the usermode application is no longer there. The usermode application does not provide any threads for processing of the operating system requests. Maybe trendmicro blocks those requests after the start of content.exe

**Comment by sdecugis** [comment]
@liryna: OK thank you. I hope this helps to identify a root cause that might also explain the behavior I am seeing.

@marinkobabic: thank you for the analysis. I am using 1 thread. What is unclear to me is that the user-space callback seems *not* being called from kernel, then a timeout occurs. Could it be the same deadlock issue #345 that generates this behavior maybe ?

**Comment by marinkobabic** [comment]
Yes it's possible. If you use more than one thread you should not face this problem or let's say the chance is small. But anyway you have to test with one thread and the deadlock fix included.

**Comment by sdecugis** [comment]
Thanks. The reason I used 1 thread only is that I have seen some race conditions in user-space previously, and I did not experience any performance penalty using a single thread so far. So using 2 threads seems more like a dirty workaround than a solution to me ^^.

I unfortunately don t have an environment to rebuild a kernel driver myself, so I either need to wait for 1.0.1 with that deadlock fix, or I need some kind soul to build a test release for me :)

Kind regards,
Sebastien.

**Comment by Rondom** [comment]
Every commit is automatically build using Appveyor continuous integration.
[Latest build (as of this writing)](https://ci.appveyor.com/project/Maxhy/dokany/build/job/ykixe75gf7i8ioq9/artifacts).

The builds are not signed by a trusted certificate, i.e. you need to trust the certificate and enable testsigning mode: https://github.com/dokan-dev/dokany/wiki/Build#signing

**Comment by sdecugis** [comment]
Understood, thank you. I will give this a try.

**Comment by Liryna** [comment]
@sdecugis I retried to make a test with the RC3 but I am really unable to get this work at a normal speed 😢 

I agree with @marinkobabic , since you are using a single thread, it explained the timeout dismount if your thread has hit the deadlock (now fixed) 👍

**Comment by sdecugis** [comment]
Ok thank you for trying. I am still waiting for feedback on the result with the new driver... Hopefully the problem will be gone and that will confirm the deadlock theory.

**Comment by sdecugis** [comment]
We have confirmed that the freeze we had sometimes with mirror.exe is gone with the latest changes in repository; however it seems the issue persists with TrendMicro 11... I will try to obtain more logs.
